--- a/drivers/of/property.c
+++ b/drivers/of/property.c
@@ -1079,8 +1079,14 @@ static void of_link_to_phandle(struct de
 {
 	struct device_node *tmp_np = of_node_get(sup_np);
 
-	/* Check that sup_np and its ancestors are available. */
+	/* Check that sup_np or its ancestors are available, */
+	/* and that sup_np is not the consumer or a child of the consumer. */
 	while (tmp_np) {
+		if (tmp_np == con_np) {
+			of_node_put(tmp_np);
+			return;
+		}
+
 		if (of_fwnode_handle(tmp_np)->dev) {
 			of_node_put(tmp_np);
 			break;
