--- a/build-aux/bootstrap
+++ b/build-aux/bootstrap
@@ -309,11 +309,11 @@ p
 q'
 
 get_version() {
-  app=$1
+  app="$1"
 
-  $app --version >/dev/null 2>&1 || { $app --version; return 1; }
+  "$app" --version >/dev/null 2>&1 || { "$app" --version; return 1; }
 
-  $app --version 2>&1 | sed -n "$get_version_sed"
+  "$app" --version 2>&1 | sed -n "$get_version_sed"
 }
 
 check_versions() {
@@ -329,7 +329,7 @@ check_versions() {
       $check_git || continue
     fi
     # Honor $APP variables ($TAR, $AUTOCONF, etc.)
-    appvar=$(echo $app | LC_ALL=C tr '[a-z]-' '[A-Z]_')
+    appvar=$(echo "$app" | LC_ALL=C tr '[a-z]-' '[A-Z]_')
     test "$appvar" = TAR && appvar=AMTAR
     case $appvar in
         GZIP) ;; # Do not use $GZIP:  it contains gzip options.
@@ -366,13 +366,13 @@ check_versions() {
     if [ "$req_ver" = "-" ]; then
       # Merely require app to exist; not all prereq apps are well-behaved
       # so we have to rely on $? rather than get_version.
-      if ! check_exists --verbose $app; then
+      if ! check_exists --verbose "$app"; then
         warn_ "Error: '$app' not found"
         ret=1
       fi
     else
       # Require app to produce a new enough version string.
-      inst_ver=$(get_version $app)
+      inst_ver=$(get_version "$app")
       if [ ! "$inst_ver" ]; then
         warn_ "Error: '$app' not found"
         ret=1
@@ -1135,7 +1135,7 @@ autogen()
   # two just-pre-run programs.
 
   # Import from gettext.
-  with_gettext=yes
+  with_gettext=no
   grep '^[	 ]*AM_GNU_GETTEXT_VERSION(' configure.ac >/dev/null || \
       with_gettext=no
 
