--- a/drivers/tty/serial/8250/8250_port.c
+++ b/drivers/tty/serial/8250/8250_port.c
@@ -2081,7 +2081,10 @@ static void wait_for_lsr(struct uart_825
 {
 	unsigned int status, tmout = 10000;
 
-	/* Wait up to 10ms for the character(s) to be sent. */
+	/* Wait for a time relative to buffer size and baud */
+	if (up->fifo_size && up->port.frame_time)
+		tmout = up->fifo_size * (up->port.frame_time / NSEC_PER_USEC);
+
 	for (;;) {
 		status = serial_in(up, UART_LSR);
 
